pkg/apis/resource/validation/validation.go

```diff
@@ -138,31 +169,81 @@ func gatherAllocatedDevices(allocationResult *resource.DeviceAllocationResult) s
 
 func validateDeviceRequest(request resource.DeviceRequest, fldPath *field.Path, stored bool) field.ErrorList {
 	allErrs := validateRequestName(request.Name, fldPath.Child("name"))
-	if request.DeviceClassName == "" {
-		allErrs = append(allErrs, field.Required(fldPath.Child("deviceClassName"), ""))
-	} else {
-		allErrs = append(allErrs, validateDeviceClassName(request.DeviceClassName, fldPath.Child("deviceClassName"))...)
+	if len(request.FirstAvailable) > 0 {
```

Comment:
As with defaults, I would expect to see this look more like:

```
if DeviceClassName == "" && len(request.FirstAvailable) == 0 {
    field.Required("one of ... must be specified")
} else if DeviceClassName != "" && len(request.FirstAvailable) > 0 {
    field.Invalid("one of ...")
} else if DeviceClassName != "" {
    // handle "exact"
} else if len(request.FirstAvailable) > 0 {
    // handle prioritized
}
```

Is there some reason that doesn't work?  We don';t need  multiple mutually-exclusive errors, I think.

Reply:
Yeah, this is clearer. I have updated the code.
We will also update this when we add the v1beta2 API.